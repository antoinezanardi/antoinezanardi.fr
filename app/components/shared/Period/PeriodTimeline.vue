<template>
  <div
    id="period-timeline"
    class="d-flex flex-column h-100"
  >
    <a
      id="timeline-logo"
      :href="url"
      rel="noopener noreferrer"
      target="_blank"
    >
      <NuxtImg
        :alt="imageAlt"
        class="logo timeline-logo-image white-logo"
        format="webp"
        :src="`/images/logos/${image}`"
      />
    </a>

    <hr class="my-3">

    <div class="d-flex flex-column flex-grow-1">
      <div
        id="period-dates"
        :aria-label="periodDatesAriaLabel"
        class="align-items-center d-flex flex-column flex-grow-1 justify-content-center"
        role="region"
      >
        <div
          id="finished-at-date"
          class="period-date"
        >
          {{ formattedFinishedAt }}
        </div>

        <WrappedFontAwesomeIcon
          v-if="doesPeriodHaveTwoDates"
          id="arrow-up-icon"
          classes="my-3"
          icon="fa-arrow-up"
          icon-color="#00000"
          size="2x"
        />

        <div
          v-if="formattedStartedAt"
          id="started-at-date"
          class="period-date"
        >
          {{ formattedStartedAt }}
        </div>
      </div>

      <div
        v-if="formattedPeriod"
        id="period-label"
        class="align-items-center d-flex flex-column"
      >
        <hr class="w-100">

        <div
          class="font-italic font-weight-bold small"
        >
          {{ formattedPeriod }}
        </div>
      </div>
    </div>
  </div>
</template>

<script lang="ts" setup>
import type { PeriodTimelineProps } from "~/components/shared/Period/period-timeline.types";
import { useDates } from "~/composables/useDates";
import WrappedFontAwesomeIcon from "~/components/shared/Icons/WrappedFontAwesomeIcon/WrappedFontAwesomeIcon.vue";
import { Period } from "~/models/period/period.class";

const props = withDefaults(defineProps<PeriodTimelineProps>(), {
  doesShowYearOnly: false,
  startedAt: undefined,
  finishedAt: undefined,
});

const { t } = useI18n();
const { getMonthFullName } = useDates();

const formattedPeriod = computed<string>(() => {
  if (!props.startedAt) {
    return "";
  }
  const finishedAt = props.finishedAt ?? new Date();
  const period = new Period(props.startedAt, finishedAt);
  let periodText = "";
  if (period.year) {
    periodText += t("shared.years", { yearCount: period.year }, period.year);
  }
  if (!props.doesShowYearOnly) {
    if (period.month && period.year) {
      periodText += ` ${t("shared.and")} `;
    }
    if (period.month) {
      periodText += t("shared.months", { monthCount: period.month });
    }
  }
  return `( ${periodText} )`;
});

const formattedStartedAt = computed<string>(() => {
  const { startedAt, doesShowYearOnly } = props;
  if (!startedAt) {
    return "";
  }
  const startedAtMonth = getMonthFullName(startedAt);
  const startedAtYear = startedAt.getFullYear().toString();

  return doesShowYearOnly ? startedAtYear : `${startedAtMonth} ${startedAtYear}`;
});

const formattedFinishedAt = computed<string>(() => {
  const { finishedAt, doesShowYearOnly } = props;
  if (finishedAt) {
    const finishedAtYear = finishedAt.getFullYear().toString();

    return doesShowYearOnly ? finishedAtYear : `${getMonthFullName(finishedAt)} ${finishedAtYear}`;
  }
  return t("shared.today");
});

const doesPeriodHaveTwoDates = computed<boolean>(() => !!formattedStartedAt.value);

const periodDatesAriaLabel = computed<string>(() => {
  const { startedAt } = props;
  if (startedAt) {
    return t("PeriodTimeline.periodDatesAriaLabel", {
      startedAt: formattedStartedAt.value,
      finishedAt: formattedFinishedAt.value,
    });
  }
  return t("PeriodTimeline.obtainedAtAriaLabel", {
    obtainedAt: formattedFinishedAt.value,
  });
});
</script>

<style lang="scss" scoped>
.period-date {
  font-size: 1.25rem;
}
.logo {
  max-width: 180px;
  max-height: 75px;
  filter: brightness(0) invert(1);
}
</style>