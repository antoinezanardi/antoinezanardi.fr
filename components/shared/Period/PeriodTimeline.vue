<template>
  <div id="period-timeline" class="d-flex flex-column h-100">
    <a :href="url" target="_blank" rel="noopener noreferrer">
      <img
        class="white-logo logo"
        :alt="image"
        :src="`assets/images/${image}`"
      />
    </a>
    <hr class="my-3"/>
    <div class="d-flex flex-column flex-grow-1">
      <div id="period-dates" class="d-flex flex-column flex-grow-1 justify-content-center align-items-center">
        <div class="period-date" v-html="formattedFinishedAt"/>
        <i v-if="doesPeriodHaveTwoDates" class="fa fa-arrow-up fa-2x my-3"/>
        <div v-if="formattedStartedAt" class="period-date" v-html="formattedStartedAt"/>
      </div>
      <div v-if="formattedPeriod" id="period-label" class="d-flex flex-column align-items-center">
        <hr class="w-100"/>
        <div class="font-weight-bold font-italic small" v-html="formattedPeriod"/>
      </div>
    </div>
  </div>
</template>

<script lang="ts" setup>
import { useI18n } from "vue-i18n";
import { computed } from "#imports";
import { useDates } from "~/composables/useDates";
import { Period } from "~/models/Period";

const props = withDefaults(defineProps<{
  url: string;
  image: string;
  doesShowYearOnly?: boolean;
  startedAt?: Date;
  finishedAt?: Date;
}>(), {
  doesShowYearOnly: false,
  startedAt: undefined,
  finishedAt: undefined,
});

const { t } = useI18n();
const { getMonthFullName } = useDates();

const formattedPeriod = computed<string | undefined>(() => {
  if (!props.startedAt) {
    return undefined;
  }
  const finishedAt = props.finishedAt ?? new Date();
  const period = new Period(props.startedAt, finishedAt);
  let periodText = "";
  if (period.year) {
    periodText += t("shared.years", { yearCount: period.year }, period.year);
  }
  if (!props.doesShowYearOnly) {
    if (period.month && period.year) {
      periodText += ` ${t("shared.and")} `;
    }
    if (period.month) {
      periodText += t("shared.months", { monthCount: period.month });
    }
  }
  return `( ${periodText} )`;
});
const formattedStartedAt = computed<string | undefined>(() => {
  const { startedAt, doesShowYearOnly } = props;
  if (!startedAt) {
    return undefined;
  }
  const startedAtMonth = getMonthFullName(startedAt);
  const startedAtYear = startedAt.getFullYear().toString();
  return doesShowYearOnly ? startedAtYear : `${startedAtMonth} ${startedAtYear}`;
});
const formattedFinishedAt = computed<string>(() => {
  const { finishedAt, doesShowYearOnly } = props;
  if (finishedAt) {
    const finishedAtYear = finishedAt.getFullYear().toString();
    return doesShowYearOnly ? finishedAtYear : `${getMonthFullName(finishedAt)} ${finishedAtYear}`;
  }
  return t("shared.today");
});
const doesPeriodHaveTwoDates = computed<boolean>(() => formattedStartedAt.value !== undefined);
</script>

<style lang="scss" scoped>
.period-date {
  font-size: 1.25rem;
}
.logo {
  max-width: 180px;
  max-height: 75px;
  filter: brightness(0) invert(1);
}
</style>