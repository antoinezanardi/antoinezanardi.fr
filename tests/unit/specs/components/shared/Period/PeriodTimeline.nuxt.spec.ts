import type { mount } from "@vue/test-utils";
import type { ComponentMountingOptions } from "@vue/test-utils/dist/mount";

import type { NuxtImg } from "#components";
import PeriodTimeline from "~/components/shared/Period/PeriodTimeline.vue";
import { mountSuspendedComponent } from "@tests/unit/utils/helpers/mount.helpers";
import type { PeriodTimelineProps } from "~/components/shared/Period/period-timeline.types";
import type WrappedFontAwesomeIcon from "~/components/shared/Icons/WrappedFontAwesomeIcon/WrappedFontAwesomeIcon.vue";

describe("Period Timeline Component", () => {
  const defaultProps: PeriodTimelineProps = {
    url: "url",
    image: "image",
    imageAlt: "imageAlt",
  };
  let wrapper: ReturnType<typeof mount<typeof PeriodTimeline>>;

  async function mountPeriodTimelineComponent(options: ComponentMountingOptions<typeof PeriodTimeline> = {}): Promise<ReturnType<typeof mount<typeof PeriodTimeline>>> {
    return mountSuspendedComponent(PeriodTimeline, {
      props: defaultProps,
      ...options,
    });
  }

  beforeEach(async() => {
    wrapper = await mountPeriodTimelineComponent();
  });

  it("should match snapshot when rendered.", () => {
    expect(wrapper).toBeTruthy();
    expect(wrapper.html()).toMatchSnapshot();
  });

  describe("Logo", () => {
    it("should set logo url to anchor when rendered.", () => {
      const logo = wrapper.find<HTMLAnchorElement>("#timeline-logo");

      expect(logo.attributes("href")).toBe("url");
    });

    it("should set logo image source when rendered.", () => {
      const logoImg = wrapper.findComponent<typeof NuxtImg>(".timeline-logo-image");

      expect(logoImg.props("src")).toBe("/images/logos/image");
    });
  });

  describe("Periods", () => {
    describe("Finished at date", () => {
      it("should display finished at today formatted label when finished at is not provided in props.", () => {
        const finishedAt = wrapper.find<HTMLDivElement>("#finished-at-date");

        expect(finishedAt.text()).toBe("shared.today");
      });

      it("should display finished at formatted label when finished at is provided in props.", async() => {
        wrapper = await mountPeriodTimelineComponent({
          props: {
            ...defaultProps,
            finishedAt: new Date("2022-01-01"),
          },
        });
        const finishedAt = wrapper.find<HTMLDivElement>("#finished-at-date");

        expect(finishedAt.text()).toBe("janvier 2022");
      });

      it("should only display finished at year when showOnlyYear is true in props.", async() => {
        wrapper = await mountPeriodTimelineComponent({
          props: {
            ...defaultProps,
            finishedAt: new Date("2022-01-01"),
            doesShowYearOnly: true,
          },
        });
        const finishedAt = wrapper.find<HTMLDivElement>("#finished-at-date");

        expect(finishedAt.text()).toBe("2022");
      });
    });

    describe("Arrow up icon", () => {
      it("should not display arrow up icon when period doesn't have two dates.", () => {
        const arrowUp = wrapper.findComponent<typeof WrappedFontAwesomeIcon>("#arrow-up-icon");

        expect(arrowUp.exists()).toBeFalsy();
      });

      it("should display arrow up icon when period has two dates.", async() => {
        wrapper = await mountPeriodTimelineComponent({
          props: {
            ...defaultProps,
            startedAt: new Date("2022-01-01"),
            finishedAt: new Date("2022-02-01"),
          },
        });
        const arrowUp = wrapper.findComponent<typeof WrappedFontAwesomeIcon>("#arrow-up-icon");

        expect(arrowUp.exists()).toBeTruthy();
      });
    });

    describe("Started at date", () => {
      it("should not display started at date when it is not provided in props.", () => {
        const startedAt = wrapper.find<HTMLDivElement>("#started-at-date");
        const periodDates = wrapper.find<HTMLDivElement>("#period-dates");

        expect(startedAt.exists()).toBeFalsy();
        expect((wrapper.vm as unknown as { formattedStartedAt: string }).formattedStartedAt).toBe("");
        expect(periodDates.text()).toBe("shared.today");
      });

      it("should match snapshot when started at date is provided in props.", async() => {
        wrapper = await mountPeriodTimelineComponent({
          props: {
            ...defaultProps,
            startedAt: new Date("2022-01-01"),
          },
        });

        expect(wrapper.html()).toMatchSnapshot();
      });

      it("should display started at formatted label when started at is provided in props.", async() => {
        wrapper = await mountPeriodTimelineComponent({
          props: {
            ...defaultProps,
            startedAt: new Date("2022-01-01"),
          },
        });
        const startedAt = wrapper.find<HTMLDivElement>("#started-at-date");

        expect(startedAt.text()).toBe("janvier 2022");
      });

      it("should only display started at year when showOnlyYear is true in props.", async() => {
        wrapper = await mountPeriodTimelineComponent({
          props: {
            ...defaultProps,
            startedAt: new Date("2022-01-01"),
            doesShowYearOnly: true,
          },
        });
        const startedAt = wrapper.find<HTMLDivElement>("#started-at-date");

        expect(startedAt.text()).toBe("2022");
      });
    });

    describe("Period", () => {
      it("should display period formatted label when started at and finished at are provided in props.", async() => {
        wrapper = await mountPeriodTimelineComponent({
          props: {
            ...defaultProps,
            startedAt: new Date("2022-01-01"),
            finishedAt: new Date("2024-03-01"),
          },
        });
        const period = wrapper.find<HTMLDivElement>("#period-label");

        expect(period.text()).toBe("( shared.years, {\"yearCount\":2}, 2 shared.and shared.months, {\"monthCount\":3} )");
      });

      it("should match snapshot when started at and finished at are provided in props.", async() => {
        wrapper = await mountPeriodTimelineComponent({
          props: {
            ...defaultProps,
            startedAt: new Date("2022-01-01"),
            finishedAt: new Date("2024-03-01"),
          },
        });

        expect(wrapper.html()).toMatchSnapshot();
      });

      it("should not display period formatted label when started at and finished at are not provided in props.", () => {
        const period = wrapper.find<HTMLDivElement>("#period-label");

        expect(period.exists()).toBeFalsy();
      });

      it("should not display period formatted label when started at is not provided in props.", async() => {
        wrapper = await mountPeriodTimelineComponent({
          props: {
            ...defaultProps,
            finishedAt: new Date("2024-03-01"),
          },
        });
        const period = wrapper.find<HTMLDivElement>("#period-label");

        expect(period.exists()).toBeFalsy();
        expect((wrapper.vm as unknown as { formattedPeriod: string }).formattedPeriod).toBe("");
      });

      it("should display only years when started at and finished at are exactly one year apart.", async() => {
        wrapper = await mountPeriodTimelineComponent({
          props: {
            ...defaultProps,
            startedAt: new Date("2022-01-01"),
            finishedAt: new Date("2022-12-31"),
          },
        });
        const period = wrapper.find<HTMLDivElement>("#period-label");

        expect(period.text()).toBe("( shared.years, {\"yearCount\":1}, 1 )");
      });

      it("should display only months when started at and finished at are exactly one month apart.", async() => {
        wrapper = await mountPeriodTimelineComponent({
          props: {
            ...defaultProps,
            startedAt: new Date("2022-01-01"),
            finishedAt: new Date("2022-01-31"),
          },
        });
        const period = wrapper.find<HTMLDivElement>("#period-label");

        expect(period.text()).toBe("( shared.months, {\"monthCount\":1} )");
      });

      it("should display years only when does show year only is true.", async() => {
        wrapper = await mountPeriodTimelineComponent({
          props: {
            ...defaultProps,
            startedAt: new Date("2022-01-01"),
            finishedAt: new Date("2024-03-01"),
            doesShowYearOnly: true,
          },
        });
        const period = wrapper.find<HTMLDivElement>("#period-label");

        expect(period.text()).toBe("( shared.years, {\"yearCount\":2}, 2 )");
      });
    });
  });
});