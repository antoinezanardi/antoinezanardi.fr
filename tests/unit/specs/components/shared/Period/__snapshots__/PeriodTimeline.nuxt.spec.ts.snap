// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Period Timeline Component > Periods > Period > should match snapshot when started at and finished at are provided in props. 1`] = `
"<div data-v-a56b5921="" id="period-timeline" class="d-flex flex-column h-100"><a data-v-a56b5921="" id="timeline-logo" href="url" rel="noopener noreferrer" target="_blank">
    <nuxt-img-stub data-v-a56b5921="" alt="imageAlt" format="webp" src="/images/logos/image" preload="false" crossorigin="false" placeholder="false" custom="false" class="logo timeline-logo-image white-logo"></nuxt-img-stub>
  </a>
  <hr data-v-a56b5921="" class="my-3">
  <div data-v-a56b5921="" class="d-flex flex-column flex-grow-1">
    <div data-v-a56b5921="" id="period-dates" aria-label="PeriodTimeline.periodDatesAriaLabel, {&quot;startedAt&quot;:&quot;janvier 2022&quot;,&quot;finishedAt&quot;:&quot;mars 2024&quot;}" class="align-items-center d-flex flex-column flex-grow-1 justify-content-center" role="region">
      <div data-v-a56b5921="" id="finished-at-date" class="period-date">mars 2024</div>
      <wrapped-font-awesome-icon-stub data-v-a56b5921="" icon="fa-arrow-up" iconcolor="#00000" size="2x" classes="my-3" id="arrow-up-icon"></wrapped-font-awesome-icon-stub>
      <div data-v-a56b5921="" id="started-at-date" class="period-date">janvier 2022</div>
    </div>
    <div data-v-a56b5921="" id="period-label" class="align-items-center d-flex flex-column">
      <hr data-v-a56b5921="" class="w-100">
      <div data-v-a56b5921="" class="font-italic font-weight-bold small">( shared.years, {"yearCount":2}, 2 shared.and shared.months, {"monthCount":3} )</div>
    </div>
  </div>
</div>"
`;

exports[`Period Timeline Component > Periods > Started at date > should match snapshot when started at date is provided in props. 1`] = `
"<div data-v-a56b5921="" id="period-timeline" class="d-flex flex-column h-100"><a data-v-a56b5921="" id="timeline-logo" href="url" rel="noopener noreferrer" target="_blank">
    <nuxt-img-stub data-v-a56b5921="" alt="imageAlt" format="webp" src="/images/logos/image" preload="false" crossorigin="false" placeholder="false" custom="false" class="logo timeline-logo-image white-logo"></nuxt-img-stub>
  </a>
  <hr data-v-a56b5921="" class="my-3">
  <div data-v-a56b5921="" class="d-flex flex-column flex-grow-1">
    <div data-v-a56b5921="" id="period-dates" aria-label="shared.today" class="align-items-center d-flex flex-column flex-grow-1 justify-content-center" role="region">
      <div data-v-a56b5921="" id="finished-at-date" class="period-date">shared.today</div>
      <wrapped-font-awesome-icon-stub data-v-a56b5921="" icon="fa-arrow-up" iconcolor="#00000" size="2x" classes="my-3" id="arrow-up-icon"></wrapped-font-awesome-icon-stub>
      <div data-v-a56b5921="" id="started-at-date" class="period-date">janvier 2022</div>
    </div>
    <div data-v-a56b5921="" id="period-label" class="align-items-center d-flex flex-column">
      <hr data-v-a56b5921="" class="w-100">
      <div data-v-a56b5921="" class="font-italic font-weight-bold small">( shared.months, {"monthCount":1} )</div>
    </div>
  </div>
</div>"
`;

exports[`Period Timeline Component > should match snapshot when rendered. 1`] = `
"<div data-v-a56b5921="" id="period-timeline" class="d-flex flex-column h-100"><a data-v-a56b5921="" id="timeline-logo" href="url" rel="noopener noreferrer" target="_blank">
    <nuxt-img-stub data-v-a56b5921="" alt="imageAlt" format="webp" src="/images/logos/image" preload="false" crossorigin="false" placeholder="false" custom="false" class="logo timeline-logo-image white-logo"></nuxt-img-stub>
  </a>
  <hr data-v-a56b5921="" class="my-3">
  <div data-v-a56b5921="" class="d-flex flex-column flex-grow-1">
    <div data-v-a56b5921="" id="period-dates" aria-label="shared.today" class="align-items-center d-flex flex-column flex-grow-1 justify-content-center" role="region">
      <div data-v-a56b5921="" id="finished-at-date" class="period-date">shared.today</div>
      <!--v-if-->
      <!--v-if-->
    </div>
    <!--v-if-->
  </div>
</div>"
`;
